File: 023/euler_23.hs
import Data.List

divisors :: (Integral a) => a -> [a]
divisors n = foldl (\acc x -> if n `mod` x == 0 then x : acc else acc) [] [1..n-1]

abundant :: (Integral a) => a -> Bool
abundant n = n < sum (divisors n)

abundantNumbers :: (Integral a) => [a]
abundantNumbers = filter abundant [1..]

sumsOfTwoAbundantNumbers :: (Integral a) => [a]
sumsOfTwoAbundantNumbers = [x+y | x <- abundantNumbers, y <- abundantNumbers]

euler23 :: (Integral a) => a
euler23 =
    let upperLimit = 28123
        notSumsofTwoAbundantNumbers = [1..upperLimit] \\ (takeWhile (<=upperLimit) sumsOfTwoAbundantNumbers)
    in sum notSumsofTwoAbundantNumbers

main = putStr $ show euler23

File: 023/euler_23.py
#!/usr/bin/env python

from functools import lru_cache
from math import sqrt


@lru_cache(maxsize=None)
def divisors(x):
    result = set([1])
    for d in range(int(sqrt(x)), 1, -1):
        if x % d == 0 and d not in result:
            result.add(d)
            _d = int(x / d)
            result.add(_d)
            _divs = divisors(_d)
            result.update(_divs)
            for a in _divs.difference(set([1])):
                result.add(int(x / a))
    return result


def abundant(x):
    return sum(divisors(x)) > x


def abundant_numbers(limit):
    if limit < 12:
        return None
    return set(filter(abundant, range(12, limit)))


def sums_of_two_abundant_numbers(limit):
    return set((x + y for x in abundant_numbers(limit) for y in abundant_numbers(limit) if x + y <= limit))


N = 28123 + 1
print(sum(set(range(1, N)).difference(sums_of_two_abundant_numbers(N))))
